bootstrap/041_quotas.sql

```diff
@@ -1,51 +1,105 @@
+-- This has the limitation that if there are more than 10K queries run in a two-hour window, we will miss some.
+CREATE OR REPLACE FUNCTION INTERNAL.GET_QUERY_HISTORY_FUNC_START_RANGE(t TIMESTAMP)
+RETURNS TIMESTAMP
+AS
+$$
+    timestampadd('hour', -1, date_trunc('hour', t))
+$$;
+
+
+CREATE OR REPLACE PROCEDURE INTERNAL.REFRESH_HOURLY_QUERY_USAGE_SQL(now TIMESTAMP)
+RETURNS STRING
+LANGUAGE SQL
+AS
 BEGIN
-    let s string := $$
--- TODO the table func caps out at 10k rows which could miss queries for busy accounts.
-with todays_queries as(
-    select
-        total_elapsed_time,
-        credits_used_cloud_services,
-        warehouse_size,
-        user_name,
-        role_name
-    from table(information_schema.query_history(
-        RESULT_LIMIT => 10000,
-        END_TIME_RANGE_START => date_trunc('day', current_timestamp()),
-        END_TIME_RANGE_END => current_timestamp()))
-),
-costed_queries as (
-    select
-        greatest(0,total_elapsed_time) * internal.warehouse_credits_per_milli(warehouse_size) + credits_used_cloud_services as credits_used,
-        user_name,
-        role_name
-    from todays_queries
-),
--- Is there a way to collapse user_usage and role_usage together? They're essentially the same query.
-user_usage as (
-    select user_name as name, sum(credits_used) as usage_map
-    from costed_queries
-    group by user_name
-),
-role_usage as (
-    select role_name as name, sum(credits_used) as usage_map
-    from costed_queries
-    group by role_name
-),
-usage_aggr as (
-    select 'users' as kind, object_agg(name, usage_map) as credits_used from user_usage union all select 'roles', object_agg(name, usage_map) from role_usage
-)
--- Aggregate into a final map {'users': {..}, 'roles': {..}}
-select object_agg(kind, credits_used) as daily_quota from usage_aggr;
+    let sql string := $$
+DECLARE
+    start_time timestamp default current_timestamp();
+    qh_func_start_time timestamp default internal.get_query_history_func_start_range('$$ || :now || $$');
+BEGIN
+    BEGIN TRANSACTION;
+
+    -- Read account_usage.query_history and summarize by user/role and write to the table
+    MERGE INTO internal.aggregated_hourly_quota q
+    USING (
+        with todays_queries as (
+            select total_elapsed_time,
+                   credits_used_cloud_services,
+                   warehouse_size,
+                   user_name,
+                   role_name,
+                   end_time
+            from table (information_schema.query_history(
+                   RESULT_LIMIT => 10000,
+                   -- TODO the function demands TIMESTAMP_LTZ. What does that actually do to our TIMESTAMP_NTZ?
+                   END_TIME_RANGE_START => TO_TIMESTAMP_LTZ(:qh_func_start_time),
+                   END_TIME_RANGE_END => current_timestamp()))
+        ), costed_queries as (
+            select
+                greatest(0, total_elapsed_time) * internal.warehouse_credits_per_milli(warehouse_size) +
```

Comment:
I think this also needs to look at warehouse id and you should use the function that does both.

Reply:
thanks! I had missed these functions. Makes sense.

Reply:
I propose that we exclude serverless tasks from this calculation and document as such. 

Let's use the already defined "is_serverless_task" function to decide if it should be considered.

Reply:
> I propose that we exclude serverless tasks from this calculation and document as such.

SGTM

> Let's use the already defined "is_serverless_task" function to decide if it should be considered.

Looks like INFORMATION_SCHEMA.QUERY_HISTORY does not return the `WAREHOUSE_ID` column that ACCOUNT_USAGE.QUERY_HISTORY has, and that `INTERNAL.IS_SERVERLESS_WAREHOUSE(NUMBER)` relies on.

Cross-referencing some queries across the table function and the view, we check the prefix of the `WAREHOUSE_NAME` as another way to check this. I went looking for some other functions which might help determine this more reliably but didn't find anything.

I've introduced a new UDF with the same name that accepts a `WAREHOUSE_NAME` instead of the `WAREHOUSE_ID`